### ISSUE TITLE
Refresh Screen Provided By `curses.wrapper` Causes Seg Fault (macOS, xcode 15 Apple supplied ncurses 6.0 breakage)

### ISSUE THREAD

**Issue by timway** [python/cpython#109617]
# Crash report

### What happened?

MacOS began pushing out updates to XCode Command Line Tools to install 15.0 recently. Upon updating I began having issues with `curses`. This happens with the Python provided by Apple. I'm not aware of the best way to communicate this issue to Apple, hopefully someone here knows who to ping or is watching.

Save the below as `curses-segfault.py`:

```python
import curses

def main(stdscr):
    stdscr.refresh()

curses.wrapper(main)
```

Run the script in `zsh` in MacOS Terminal via:

```zsh
/usr/bin/python3 curses-segfault.py
```

An easy way to see if you got the update is via the terminal by running:

```zsh
softwareupdate --history
```

### CPython versions tested on:

3.9

### Operating systems tested on:

macOS

### Output from running 'python -VV' on the command line:

python3 -VV Python 3.9.6 (default, Aug 11 2023, 19:44:49)  [Clang 15.0.0 (clang-1500.0.40.1)]

### Linked PRs
* gh-111258

**Comment by brandtbucher** [comment]
I've been encountering this on `main` too after updating (the `test_curses` tests in our suite seem to trigger it). One *really* annoying side-effect is that it breaks your terminal after the crash.

I'm not a `curses` expert, but I'm not sure that this is something we can fix on our end, so this should probably be closed. It might be worth disabling the crashy tests if we can detect the broken curses version, though?

**Comment by brandtbucher** [comment]
CC @Yhg1s as a curses expert.

**Comment by brandtbucher** [comment]
The affected tests appear to be `test_create_windows`, `test_move_cursor`, `test_output_character`, `test_refresh`, and `test_refresh_control`.

**Comment by ned-deily** [comment]
FWIW this problem should not arise when using a python from a macOS python.org installer or MacPorts as they supply their own version of ncurses and do not use the system version.

**Comment by robgilreath** [comment]
> FWIW this problem should not arise when using a python from a macOS python.org installer or MacPorts as they supply their own version of ncurses and do not use the system version.

I can confirm this. I was only able to reproduce it for @timway when using the system version of python.

**Comment by brandtbucher** [comment]
Downgrading from Xcode 15.0 (`curses.ncurses_version(major=6, minor=0, patch=20150808)`) to 14.3 (`curses.ncurses_version(major=5, minor=7, patch=20081102)`) fixes it for me.

**Comment by gpshead** [comment]
It sounds like they shipped a potentially broken version of ncurses embedded within xcode 15?

We've been building and linking with ncurses >= 6 in the rest of the world since 2015 without this problem...

**Comment by gpshead** [comment]
It is really strange behavior on Apple's part to "upgrade" to providing 6.0 from 2015 in their 2023 toolchain when ncurses 6.4 came out last year.

We're at the point where it'd be fine for CPython to simply drop the ability to build and link against 6.0 in 3.13+ given that no supported Linux distro platform will be using it anymore by the time we release 3.13.  (RHEL 8 shipped with ncurses 6.1 making that a reasonable minimum) -- _I do not expect or encourage anyone to go through and do that to the `_curses` module code._

Recommendation: do what we do with python.org builds, never build against apple's ncurses.

If a workaround for this isn't obvious to someone with a macos debugger, it is reasonable to have a configure check blocklist the xcode 15 supplied ncurses and refuse to use it. Spending _our_ time dealing with brazenly outdated libraries embedded in someone elses toolchain that we don't ship our own builds with isn't worthwhile (it'd never end).

**Comment by timway** [comment]
Thank you all for looking into this - as next steps I've found the Feedback Assistant and the general feedback form on Apple's website. I'll work to post to those methods and provide any feedback here.

I could also look at writing a small C example and see if I can confirm if the problem is in their ncurses implementation and fully removed from Python or not.

**Comment by vstinner** [comment]
If you want to require a specific version or recommend libedit or ncurses on macOS, you can update the doc: https://docs.python.org/dev/using/configure.html

**Comment by ronaldoussoren** [comment]
Issues like this should also be reported to Apple as this is a crash with their copy of Python.

I've filed FB13196764 about this.

**Comment by gpshead** [comment]
Thanks Ronald!

**Comment by ronaldoussoren** [comment]
> FWIW this problem should not arise when using a python from a macOS python.org installer or MacPorts as they supply their own version of ncurses and do not use the system version.

Our copy of ncurses is ancient though and should be updated to a more recent version. But that's something for a different issue, and needs careful testing because we rely on the system terminfo database.

**Comment by vstinner** [comment]
Maybe the magic macOS macro ``__builtin_available(macOS ...)`` should be used for ncurses.

**Comment by vstinner** [comment]
As a fan of Python limited C API, I would prefer that Python does not use NCURSES_OPAQUE=0. While it works today, it may break tomorrow.

Another way should be found to fix the Python module.

"is_pad: returns TRUE if the window is a pad i.e., created by newpad"

If the function is not available, can we just not make it available in Python?

**Comment by ronaldoussoren** [comment]
> On 11 Oct 2023, at 17:21, Davide Rizzo ***@***.***> wrote:
> 
> ﻿
> Maybe the magic macOS macro __builtin_available(macOS ...) should be used for ncurses.
> 
> It shouldn't be necessary. It's a funny decision on Apple's side to use availability attributes on a function like is_pad. How do you even implement a fallback if the function is not available at runtime?
> 

> For ncurses, it should be enough to default to NCURSES_OPAQUE=0 which:
> 

Without having looked at the source code: I’d prefer using builtin_available instead of disabling is_pad.

**Comment by ronaldoussoren** [comment]
Sorry about the slow response, I was traveling at the time.

> I don't understand what "instead of disabling is_pad" means.

You wrote:

> The root issue is that is_pad is weakly linked and is only available in macOS 14 (and iOS 17 etc etc), but it's exposed in recent SDKs. This confuses the configure check for HAVE_CURSES_IS_PAD which will see the function, although it may be unavailable at runtime.

> The compiler warns about this with a warning: 'is_pad' is only available on macOS 14.0 or newer [-Wunguarded-availability-new].

The IMHO correct way to deal with this is to use ``is_pad`` (and other curses functions new in 14.0) only when running on 14.0 or later by using ``__builtin_available``.  This does complicate the code in ``Modules/_cursesmodule.c`` a little, but the amount of change likely can be limited by slightly restructuring the code (such as making ``py_is_pad`` a static inline function instead of a macro).   We already do this in a number of other modules, including some much more involved changes in posixmodule.c.

**Comment by vstinner** [comment]
```
if (HAVE_CURSES_IS_PAD_RUNTIME) {
    return is_pad(w);
} else {
    // ... what goes here? ...
}
```

Don't define `is_pad` macro if the function is not available. `Modules/_cursesmodule.c` can already be built without is_pad() function. Example in _curses_window_echochar_impl():

```c
#ifdef py_is_pad                        
    if (py_is_pad(self->win)) {                                          
        return PyCursesCheckERR(pechochar(self->win, ch_ | (attr_t)attr),   
                                "echochar");
    }                                                                   
    else                                                       
#endif                                                                   
        return PyCursesCheckERR(wechochar(self->win, ch_ | (attr_t)attr),
                                "echochar");       
```

Or well, just always return 0?

**Comment by vstinner** [comment]
I understand that with Xcode 15.0, the is_pad() function is **not** available, but `win->_flags & _ISPAD` can be tested if the ``NCURSES_OPAQUE`` macro is set to 0. For me, it's surprising to have to tune NCURSES_OPAQUE macro and to inspect a structure member starting with an underscore.

But apparently, this issue affects many macOS issues, it's annoying, and I should stop nitpicking. So just use NCURSES_OPAQUE=0 on macOS, and it should "just works", no?

Does someone want to propose a PR? Nobody proposed a PR so far, no?

**Comment by efalk** [comment]
Is there a workaround? I want to get back to work on my project.

**Comment by timway** [comment]
> Is there a workaround? I want to get back to work on my project.

I'm no longer able to replicate the issue with Sonoma 14.2.1 and Command Line Tools for Xcode 15.1.

Alternatively, versions from Python.org shouldn't have the issue.

**Comment by efalk** [comment]
Yes, downloading the latest directly from python.org made the problem go away. It would be nice if there were something I could do in code, other than telling customers "you need to install a new version of Python on your system"

**Comment by ned-deily** [comment]
FYI, as of the Python 3.14.0 alpha 5 pre-release and the upcoming 3.13.3 and 3.12.10 releases, python.org macOS installers now build and link with ncurses 6.5, the current stable release, and no longer with either ncurses 5.9 or with the Apple-supplied and -patched macOS system ncurses.

Since I think the original issue reported here has been addressed by [PR #111258](https://github.com/python/cpython/pull/111258) and possibly by Apple - at least there don't seem to be recent reports of people being bitten by this - let's close this issue as fixed.
